google/cloud/storage/_signing.py

```diff
@@ -542,7 +541,11 @@ def generate_signed_url_v4(
         request_timestamp = _request_timestamp
         datestamp = _request_timestamp[:8]
 
-    client_email = credentials.signer_email
+    client_email = service_account_email
+    if not access_token or not service_account_email:
+        ensure_signed_credentials(credentials)
+        client_email = credentials.signer_email
```

Comment:
Please update the docstring above to indicate that `credentials` must proviide `signer_email` only if `service_account_email` and `access_token` are not passed.

Also, this new branch (where `credentials` is ignored because `service_account_email` is passed) will require a separate unit test below.

Reply:
Docstring added, and new test cases added.